src/ctapipe/monitoring/calculator.py

```diff
@@ -0,0 +1,327 @@
+        if self.chunk_shift is None:
+            raise ValueError(
+                "chunk_shift must be set if second pass over the data is requested"
+            )
+        # Check if at least one chunk is faulty
+        if np.all(valid_chunks):
+            raise ValueError(
+                "All chunks are valid. The second pass over the data is redundant."
+            )
+        # Get the aggregator
+        aggregator = self.stats_aggregator[self.stats_aggregator_type.tel[tel_id]]
+        # Conduct a second pass over the data
+        aggregated_stats_secondpass = []
+        faulty_chunks_indices = np.where(~valid_chunks)[0]
+        for index in faulty_chunks_indices:
+            # Log information of the faulty chunks
+            self.log.info(
+                "Faulty chunk detected in the first pass at index '%s'.", index
+            )
+            # Calculate the start of the slice depending on whether the previous chunk was faulty or not
+            slice_start = (
+                aggregator.chunk_size * index
+                if index - 1 in faulty_chunks_indices
+                else aggregator.chunk_size * (index - 1)
+            )
+            # Set the start of the slice to the first element of the dl1 table if out of bound
+            # and add one ``chunk_shift``.
+            slice_start = max(0, slice_start) + self.chunk_shift
+            # Set the end of the slice to the last element of the dl1 table if out of bound
+            # and subtract one ``chunk_shift``.
+            slice_end = min(len(table) - 1, aggregator.chunk_size * (index + 2)) - (
+                self.chunk_shift - 1
+            )
+            # Slice the dl1 table according to the previously calculated start and end.
+            table_sliced = table[slice_start:slice_end]
+            # Run the stats aggregator on the sliced dl1 table with a chunk_shift
+            # to sample the period of trouble (carflashes etc.) as effectively as possible.
+            # Checking for the length of the sliced table to be greater than the ``chunk_size``
+            # since it can be smaller if the last two chunks are faulty. Note: The two last chunks
+            # can be overlapping during the first pass, so we simply ignore them if there are faulty.
+            if len(table_sliced) > aggregator.chunk_size:
+                aggregated_stats_secondpass.append(
+                    aggregator(
+                        table=table_sliced,
+                        masked_pixels_of_sample=masked_pixels_of_sample,
+                        col_name=col_name,
+                        chunk_shift=self.chunk_shift,
+                    )
+                )
+        # Stack the aggregated statistics of each faulty chunk
+        aggregated_stats_secondpass = vstack(aggregated_stats_secondpass)
+        # Detect faulty pixels with multiple instances of OutlierDetector of the second pass
+        outlier_mask_secondpass = np.zeros_like(
```

Comment:
this is the same code as in the `first_pass`. Should be factored out to a `def _find_outliers(...)`